gravitee-am-identityprovider/gravitee-am-identityprovider-azure-ad/src/main/java/io/gravitee/am/identityprovider/azure/authentication/AzureADAuthenticationProvider.java

```diff
@@ -0,0 +1,364 @@
+        return io.gravitee.am.common.oauth2.ResponseType.CODE.equals(getResponseType()) ? AuthenticationFlow.AUTHORIZATION_CODE_FLOW : AuthenticationFlow.IMPLICIT_FLOW;
+    }
+
+    @Override
+    public Maybe<User> loadUserByUsername(Authentication authentication) {
+        return authenticate(authentication)
+                .flatMap(token -> profile(token, authentication));
+    }
+
+    @Override
+    public Maybe<User> loadUserByUsername(String username) {
+        return Maybe.empty();
+    }
+
+    private Maybe<Token> authenticate(Authentication authentication) {
+        // authorization code flow, exchange code for an access token
+        // prepare body request parameters
+        final String authorizationCode = authentication.getContext().request().parameters().getFirst(configuration.getCodeParameter());
+        if (authorizationCode == null || authorizationCode.isEmpty()) {
+            LOGGER.debug("Authorization code is missing, skip authentication");
+            return Maybe.error(new BadCredentialsException("Missing authorization code"));
+        }
+
+        List<NameValuePair> urlParameters = new ArrayList<>();
+        urlParameters.add(new BasicNameValuePair(Parameters.CLIENT_ID, configuration.getClientId()));
+        urlParameters.add(new BasicNameValuePair(Parameters.CLIENT_SECRET, configuration.getClientSecret()));
+        urlParameters.add(new BasicNameValuePair(Parameters.REDIRECT_URI, (String) authentication.getContext().get(Parameters.REDIRECT_URI)));
+        urlParameters.add(new BasicNameValuePair(Parameters.CODE, authorizationCode));
+        urlParameters.add(new BasicNameValuePair(Parameters.GRANT_TYPE, "authorization_code"));
+        String bodyRequest = URLEncodedUtils.format(urlParameters);
+
+        return client.postAbs(getTokenUri())
+                .putHeader(HttpHeaders.CONTENT_LENGTH, String.valueOf(bodyRequest.length()))
+                .putHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED)
+                .rxSendBuffer(Buffer.buffer(bodyRequest))
+                .toMaybe()
+                .map(httpResponse -> {
+                    if (httpResponse.statusCode() != 200) {
+                        throw new BadCredentialsException(httpResponse.statusMessage());
+                    }
+
+                    JsonObject response = httpResponse.bodyAsJsonObject();
+                    String accessToken = response.getString(ACCESS_TOKEN_PARAMETER);
+
+                    // put the id_token in context for later use
+                    String idToken = response.getString(ID_TOKEN_PARAMETER);
+                    authentication.getContext().set(ID_TOKEN_PARAMETER, idToken);
+                    System.out.println("TOKEN: " + response);
+                    return new Token(accessToken, TokenTypeHint.ACCESS_TOKEN);
+                });
+
+    }
+
+    private Maybe<User> profile(Token token, Authentication authentication) {
```

Comment:
For Azure AD I think we can rely only on the ID Token and remove all the stuff about the UserInfo Endpoint.